### PR TITLE
Track metrics in Prometheus

### PR DESCRIPTION
**Pull request by ivanayov**
This changes adds Prometheus to the standard infra deploy and
exposes all metrics from UPDATE events from both vpp and kernel
forwarders.

The records are of type:
`rx_bytes(src_pod_name, src_pod_namespace,
         dst_pod_name, dst_pod_namespace)`
for the following kinds of metrics:
rx_bytes, tx_bytes, rx_packets, tx_packets, rx_error_packets and
tx_error_packets

Pod names and namespaces are still not exposed to cross connects
so temporary they are replaced with `netnsInode` for local src/dst
pods, `network_service_endpoint_name` for remote src/dst pods and
namespaces are temporary hardcoded.

Added to helm. Enabled with
`PROMETHEUS=1`

Signed-off-by: Ivana Yovcheva <iyovcheva@vmware.com>

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [x] Added integration testing to cover
- [x] Tested locally - nsm, icmp and crossconnect-monitor deployments with enabled metrics and prometheus
Screenshot 2019-12-05 at 17 06 32

- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by edwarnicke** [comment]
@ivanayov Is this still in progress?

**Comment by ivanayov** [comment]
@edwarnicke no, I'm going to push the integration test I added before the issue with the metrics was fixed in master + resolve the conflicts and test again.
This was paused, because of a bunch of conferences and PTO after them.

**Comment by edwarnicke** [comment]
@ivanayov Speak up if you need any help :)

**Comment by ivanayov** [comment]
thanks @edwarnicke 

The license check is failing, even though I'm using [this](https://github.com/networkservicemesh/networkservicemesh/blob/a948f02481b1b644a17cf086488a0e40715fb3c4/conf/boilerplate2.txt) one.

**Comment by edwarnicke** [comment]
You can see an example that passes the license check [here](https://github.com/networkservicemesh/networkservicemesh/blob/master/sdk/endpoint/dns.go#L1) ... try that (subbing in the VMWare Copyright).  Its a little sensitive to the [format of the header](https://github.com/networkservicemesh/networkservicemesh/blob/master/go-header.yaml#L12)

**Comment by ivanayov** [comment]
Fails provisioning in all clouds, which is the reason for failed tests.

File: k8s/conf/crossconnect-monitor-svc.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+---
+apiVersion: v1
+kind: Service
```
Comment (nickolaev) [inline k8s/conf/crossconnect-monitor-svc.yaml]: Not needed here. Helm charts shoudl be good enough

File: k8s/cmd/admission-webhook/patches.go [hunk 2]
```diff
@@ -95,10 +95,17 @@ func createDNSPatch(tuple *podSpecAndMeta, annotationValue string) (patch []patc
...
+	envVals := []corev1.EnvVar{
+		{
+			Name:  client.AnnotationEnv,
+			Value: annotationValue,
+		},
+		{
+			Name:  client.NamespaceEnv,
+			Value: namespace,
+		},
+	}
```
Comment (nickolaev) [inline k8s/cmd/admission-webhook/patches.go]: Is this needed?

File: k8s/cmd/admission-webhook/main.go [hunk 3]
```diff
@@ -84,6 +84,14 @@ func getTracerEnabled() string {
...
 }
 
+func getNamespace() string {
+	ns := os.Getenv(namespaceEnv)
+	if len(ns) == 0 {
+		ns = namespaceDefault
+	}
+	return ns
+}
+
```
Comment (nickolaev) [inline k8s/cmd/admission-webhook/main.go]: Is this needed?

File: sdk/client/const.go [hunk 4]
```diff
@@ -6,4 +6,6 @@ const (
 	AnnotationEnv = "NS_NETWORKSERVICEMESH_IO"
 	// NSMConnectRetries - Specify how many time we could retry to do a connection retrying.
 	NSMConnectRetries = "NSM_CLIENT_CONNECT_RETRIES"
+	// NamespaceEnv - NamespaceEnv is the namespace a container is running in
+	NamespaceEnv = "NSM_NAMESPACE"
```
Comment (nickolaev) [inline sdk/client/const.go]: This constant is in several places?

File: controlplane/api/metrics/helpers.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/crossconnect"
+)
+
+// Returns source and destination of the metrics
+// specified with `namespace` or `network_service_endpoint_name`
+// for local or remote source/dest respectively
+// TODO: edit to return src/dst pod names and namespaces once
+// they are available in labels
```
[the remark this answers is not in the document]
Comment (ivanayov) [reply controlplane/api/metrics/helpers.go]: No, seems I forgot it

File: k8s/cmd/admission-webhook/patches.go (hunk 2 above)
Comment (ivanayov) [reply k8s/cmd/admission-webhook/patches.go]: This is a fix for the proper k8s namespace to be exposed in crossconnect-monitor. Previously it was always `default` for the client.

File: k8s/cmd/admission-webhook/main.go (hunk 3 above)
Comment (ivanayov) [reply k8s/cmd/admission-webhook/main.go]: Same here.

File: .mk/helm.mk [hunk 6]
```diff
@@ -34,6 +34,8 @@ $(INSTALL_CHARTS):
 	--set org="${CONTAINER_REPO}",tag="${CONTAINER_TAG}" \
 	--set forwardingPlane="${FORWARDING_PLANE}" \
 	--set insecure="${INSECURE}" \
+	--set prometheus="${PROMETHEUS}" \
+	--set metricsCollectorEnabled="${METRICS_COLLECTOR_ENABLED}" \
```
[the remark this answers is not in the document]
Comment (ivanayov) [reply .mk/helm.mk]: Yes, `PROMETHEUS` is for the `crossconnect-monitor` chart, while `METRICS_COLLECTOR_ENABLED` is for `nsm` (the forwarding plane) as well. Previously there wasn't env var to enable the metrics with helm, and they are disabled by default.

File: sdk/client/const.go (hunk 4 above)
Comment (ivanayov) [reply sdk/client/const.go]: There is second in the admission-webhook, to skip depending on sdk. It's used only for exposing the namespace to the client labels (in order to be exposed to the crossconnect-monitor)

File: controlplane/api/go.mod [hunk 7]
```diff
@@ -4,6 +4,8 @@ require (
 	github.com/golang/protobuf v1.3.2
 	github.com/networkservicemesh/networkservicemesh/utils v0.2.0
 	github.com/pkg/errors v0.8.1
+	github.com/prometheus/client_golang v1.2.1
```
Comment (edwarnicke) [inline controlplane/api/go.mod]: I'm a little nervous about pulling Prometheus in as an API dependency.

File: controlplane/api/metrics/prometheus.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	prometheusHostPort := ""
+	if len(prometheusAddress) > 0 {
+		prometheusHostPort = prometheusAddress[0]
+	} else {
+		prometheusHostPort = prometheusPort
+	}
+	server := &http.Server{Addr: prometheusHostPort}
+	http.Handle("/metrics", promhttp.Handler())
+	return server
+}
```
Comment (edwarnicke) [inline controlplane/api/metrics/prometheus.go]: Could we move this out of API and into pkg?

File: go.mod [hunk 9]
```diff
@@ -51,3 +51,17 @@ replace (
...
+	github.com/pelletier/go-toml v1.6.0 // indirect
+	github.com/prometheus/client_golang v1.2.1
+	github.com/spf13/cobra v0.0.5 // indirect
+	github.com/spf13/pflag v1.0.5 // indirect
+	golang.org/x/sys v0.0.0-20191118090420-b5d5184f72d2 // indirect
+	gopkg.in/yaml.v2 v2.2.5 // indirect
+	gopkg.in/yaml.v3 v3.0.0-20191107175235-0b070bb63a18 // indirect
+	k8s.io/apimachinery v0.0.0-20191116203941-08e4eafd6d11 // indirect
+	sigs.k8s.io/kind v0.5.1 // indirect
+)
```
Comment (edwarnicke) [inline go.mod]: Why are we picking up all these extra dependencies in the root go.mod ?

**Comment by ivanayov** [comment]
Thanks for the review @edwarnicke. I addressed your comments and did some more tidy-up in the mod files. Please check if you're ok with them now.

**Comment by dtornow** [comment]
Drive-by-comment: src_pod_name, src_pod_namespace, dst_pod_name, dst_pod_namespace will be high cardinality labels, once the values are available. Have a look at [Cardinality is key](https://www.robustperception.io/cardinality-is-key)

**Comment by ivanayov** [comment]
Thank you @dtornow , I'm going to address in a follow-up, as the CI is green and takes long to run

**Pull request merged**